lib/active_merchant/billing/gateways/airwallex.rb

```diff
@@ -0,0 +1,257 @@
+      end
+
+      def setup_access_token
+        token_headers = {
+          'Content-Type' => 'application/json',
+          'x-client-id' => @client_id,
+          'x-api-key' => @client_api_key
+        }
+        response = ssl_post(build_request_url(:login), nil, token_headers)
+        @access_token = JSON.parse(response)['token']
+      end
+
+      def build_request_url(action, id = nil)
+        base_url = (test? ? test_url : live_url)
+        base_url + ENDPOINTS[action].to_s % { id: id }
+      end
+
+      def create_payment_intent(money, options = {})
+        post = {}
+        add_invoice(post, money, options)
+        post[:request_id] = options[:request_id]
+        post[:merchant_order_id] = options[:merchant_order_id]
+
+        response = commit(:setup, post)
+        response.params['id']
+      end
+
+      def add_billing(post, card, options = {})
+        return unless card_has_billing_info(card, options)
+
+        billing = post['payment_method']['card']['billing'] || {}
+        billing['email'] = options[:email] if options[:email]
+        billing['phone'] = options[:phone] if options[:phone]
+        billing['first_name'] = card.first_name
+        billing['last_name'] = card.last_name
+        billing['address'] = add_address(card, options) if card_has_address_info(card, options)
+
+        post['payment_method']['card']['billing'] = billing
+      end
+
+      def card_has_billing_info(card, options)
+        card.first_name && card.last_name
+      end
+
+      def card_has_address_info(card, options)
+        # These fields are required if address data is sent.
+        options[:address1] && options[:country]
+      end
+
+      def add_address(card, options = {})
+        address = {}
+        address[:country_code] = options[:country]
+        address[:street] = options[:address1]
+        address[:city] = options[:city] if options[:city] # required per doc, not in practice
+        address[:postcode] = options[:zip] if options[:zip]
+        address[:state] = options[:state] if options[:state]
+        address
+      end
+
+      def add_invoice(post, money, options)
+        post[:amount] = amount(money)
+        post[:currency] = (options[:currency] || currency(money))
+      end
+
+      def add_card(post, card, options = {})
+        post['payment_method'] = {
+          'type' => 'card',
+          'card' => {
+            'expiry_month' => format(card.month, :two_digits),
+            'expiry_year' => card.year.to_s,
+            'number' => card.number.to_s,
+            'name' => card.name,
+            'cvc' => card.verification_value
+          }
+        }
+        add_billing(post, card, options)
+      end
+
+      def add_payment_method_options(post, card, options = {})
+        return unless options.include?(:auto_capture) && options[:auto_capture] == false
+
+        post['payment_method_options'] = { 'card' => { 'auto_capture' => false } }
+      end
```

Comment:
The logic here is a little complex because you have to recognize that this is being called both from a `purchase` and an `authorize` transaction. I wonder if removing the guard clause and instead adding a conditional to the caller would make the relationship a little clearer, like this:

```
def purchase(money, card, options = {})
  payment_intent = create_payment_intent(money, @options)
  post = {
    'request_id' => update_request_id(@options, 'purchase'),
    'return_url' => options[:return_url]
  }
  add_card(post, card, options)
  add_payment_method_options(post, card, options) if authorization?(options)

  commit(:sale, post, payment_intent)
end

def authorization?(options)
  options.include?(:auto_capture) && options[:auto_capture] == false
end
```

Reply:
:brilliant: